### DISCUSSION TITLE
[TURBO-1043] Allow step to depend on `pnpm install`

### DISCUSSION THREAD

**Discussion by maschwenk** [vercel/turborepo#7473] [Ideas]
### Which project is this feature idea for?

Turborepo

### Describe the feature you'd like to request

Sometimes our build scripts can fail because we need to `pnpm install`. For turbo to be a one-stop shop, it would be ideal if `pnpm install` was a pre-req. Almost every turbo task has a dependency on `pnpm install`.

### Describe the solution you'd like

Turborepo knows when the lockfile has changed (I think this is part of the cache invalidation?), so it seems to follow that I should be able to define a `build` task that depends on `pnpm install` being run, and (maybe?) it should be able to be cached if the lockfile has not changed.

### Describe alternatives you've considered

🤷🏼 ?

<sub>[WEB-957](https://linear.app/vercel/issue/WEB-957/allow-step-to-depend-on-pnpm-install)</sub>

<sub>[TURBO-1143](https://linear.app/vercel/issue/TURBO-1143/[turbo-1043]-allow-step-to-depend-on-pnpm-install)</sub>

**Comment by arlyon** [comment]
Hi! We have actually been planning on injecting synthetic steps into the build graph (such as running `install` before any task) but do not have a timeline yet. It is definitely something we are working towards though. Maybe one day you can use run any project using turbo and not have to use the underlying package manager at all..

We will update this issue when we have a concrete timeline.

**Comment by jakeleventhal** [comment]
There is an easy workaround. Just create a script in the root package.json `"package-deps": "pnpm install"`

Then you can add `"//#package-deps"` to your turbo.json like so:
```json
"//#package-deps": {
  "inputs": ["**/package.json", ".npmrc", ".npmignore", "pnpm-workspace.yaml"],
  "outputs": ["pnpm-lock.yaml"]
},
"build": {
  "dependsOn": ["//#package-deps"]
}
```
also for bonus points you can add postinstall scripts to generate prisma clients or anything else like that

**Comment by maschwenk** [comment]
@jakeleventhal I did think about that but wasn't sure if there was some kind of edge case to know about. On CI we'd probably not want this just because we try to avoid installing all our dependencies, but for sure locally this seems like an easy win.

**Comment by maschwenk** [comment]
@arlyon I guess it's worth asking...do you see any issue with that approach?

**Comment by maschwenk** [comment]
@jakeleventhal

1. Our usecase would be different, we'd probably make the task `pnpm install --frozen-lockfile` and put `pnpm-lock.yaml` in the inputs, with no outputs. Most of the time we just want to be able to skip the install step because even when there's no work to do, `pnpm install --frozen-lockfile` can be slow.
2. Would having `**/package.json` in that inputs mean that `package.json`'s of your nested node_modules would invalidate the cache? My guess is your intent would be to only watch `package.json` of your own packages. Unsure of how the globbing works, does it ignore gitignored stuff?

**Comment by arlyon** [comment]
I think for this usecase the package-deps script is a good choice. I have been experimenting with using turbo and wrapping flask (or other arbitrary toolchains) and we plan on coming up with a convention for this type of task (which will be inherited by default when using one of the supported node ecosystem package managers).

To answer Q2 @maschwenk, you can use a negative glob to prevent package jsons in node_modules from being included if you wanted to be on the safe side (`!**/node_modules/**/package.json`) but they should be ignored by default anyways since we should respect your gitignore.

**Comment by andreilgeorgescu** [comment]
Has there been any progress on this feature or opinions formed after using the `package-deps` idea?
